tests/test_model_io.py

```diff
@@ -0,0 +1,127 @@
+    pend_path = ROOT + "/models/pendulum/pendulum.xml"
+    model1 = _modify_robot_float_base(pend_path)
+    combined1 = "\t".join(get_joint_names(model1))
+    assert "freejoint" in combined1
+
+    # case 2: model's first body has no joints, so add a freejoint
+    dummy_xml_string = """
+    <mujoco model="parent">
+      <worldbody>
+        <body>
+          <geom name="foo" type="box" pos="-0.2 0 0.3" size="0.5 0.3 0.1"/>
+          <site name="attachment_site" pos="1. 2. 3." quat="1. 0. 0. 1."/>
+          <body name="child" pos="1. 2. 3." quat="1. 0. 0. 1.">
+            <geom name="my_box" type="box" pos="0.5 0.25 1." size="0.1 0.2 0.3"/>
+          </body>
+        </body>
+      </worldbody>
+    </mujoco>
+    """
+    with open("_temp.xml", "w") as f:
+        f.write(dummy_xml_string)
+
+    model2 = _modify_robot_float_base("_temp.xml")
+    Path.unlink("_temp.xml")
+    combined2 = "\t".join(get_joint_names(model2))
+    assert "freejoint" in combined2
+
+    # case 3: add a freejoint to a URDF model with assets (much trickier b/c of file paths)
+    from ambersim.utils.io_utils import load_mjx_model_from_file
+
+    _, data_unfree = load_mjx_model_from_file("models/barrett_hand/bh280.urdf", force_float=False)
+    assert len(data_unfree.qpos) == 8  # 8 DOFs for the joints
+
+    _, data_free = load_mjx_model_from_file("models/barrett_hand/bh280.urdf", force_float=True)
+    assert len(data_free.qpos) == 15  # additional 7 quat states
+
+
+def test_convex_decomposition():
+    """Tests the convex decomposition util."""
+    meshfile = "models/barrett_hand/meshes/finger.obj"
+    savedir = Path("_test_dir")
+    savedir.mkdir(parents=True, exist_ok=False)
+
+    # tests that meshes are generated and saved correctly
+    decomposed_meshes = convex_decomposition_file(meshfile, quiet=True, savedir=savedir)
+    assert len(decomposed_meshes) > 0
+    assert len(list(Path(savedir).glob("*.obj"))) > 0
+
+    # tests that the decomposed meshes don't change compared to those in the mjx backend
+    # in particular, checks that both trimesh and our decompositions have very close distance queries
+    # see: https://github.com/google-deepmind/mujoco/blob/57e6940f579484adf34eebedc51279a818909f34/mjx/mujoco/mjx/_src/mesh.py#L195-L208
+    for mesh in decomposed_meshes:
+        # decomposed
+        dverts = mesh.vertices
+        dfaces = mesh.faces
+
+        # trimesh convex hull
+        tm = trimesh.Trimesh(vertices=dverts, faces=dfaces)
+        tm_convex = trimesh.convex.convex_hull(tm)
+        tverts = tm_convex.vertices
+        tfaces = tm_convex.faces
+
+        # checking that signed distance queries are functionally the same
+        coords = np.random.randn(100, 3)
+        signed_dist_d = igl.signed_distance(coords, dverts, dfaces)[0]
+        signed_dist_t = igl.signed_distance(coords, tverts, tfaces)[0]
+        assert np.allclose(signed_dist_d, signed_dist_t)
```

Comment:
@kli58 is this enough for the test case? 

In [these lines](https://github.com/google-deepmind/mujoco/blob/57e6940f579484adf34eebedc51279a818909f34/mjx/mujoco/mjx/_src/mesh.py#L195C23-L207), the `tm_convex` object gets fed into the `mjx` ecosystem with their utils, but I'm not actually testing against the `mjx` outputs, since we can't prevent it from preprocessing our input geoms anyway.

Reply:
reference to `igl`'s `signed_distance` function: https://libigl.github.io/libigl-python-bindings/igl_docs/#signed_distance